### PR TITLE
Update debugger guide

### PR DESCRIPTION
**Pull request by praveennadarajah**
## Purpose
$subject

Fixes https://github.com/ballerina-platform/ballerina-dev-website/issues/3171 https://github.com/ballerina-platform/ballerina-dev-website/issues/3365

### CODE REVIEW

File: learn/guides/visual-studio-code-extension/debugging.md [hunk 1]
```diff
@@ -156,7 +230,7 @@ Visual Studio Code allows you to debug Ballerina programs through the Ballerina
 - Strands
 - Expression Evaluation
 
->**Info** The Ballerina debugger is an open-source project and contributors are most welcome to get engaged with it via the [ballerina-lang](https://github.com/ballerina-platform/ballerina-lang) GitHub repository. If you encounter any difficulties when using this feature, feel free to create an issue on it.
+>**Info:** For more information on debugging features using VS Code, go to the [VS Code Documentation](https://code.visualstudio.com/docs/editor/debugging).
```
Comment (NipunaRanasinghe) [inline learn/guides/visual-studio-code-extension/debugging.md:233]: Can we move this info section to the end of the `Using the Debugging Features` section?

Comment (praveennadarajah) [reply learn/guides/visual-studio-code-extension/debugging.md]: Fixed with https://github.com/ballerina-platform/ballerina-dev-website/pull/3385/commits/94c36033750840488ec1bd7dc58f95dcbb463de3

**Review by NipunaRanasinghe — APPROVED** [review]

File: learn/guides/visual-studio-code-extension/debugging.md [hunk 2]
```diff
@@ -5,7 +5,7 @@ description: Describes debugging functionalities provided by Ballerina in Visual
 keywords: ballerina debugging, ballerina debug, ballerina debugger, ballerina vscode
 permalink: /learn/visual-studio-code-extension/debugging/
 active: debugging
-intro: The VS Code Ballerina extension allows you to either run your Ballerina program (without debugging) or debug them easily by launching its debugger. 
+intro: The VS Code Ballerina language extension comes with in-built debugging capabilities and provides the same debugging experience as the conventional VS Code Debugging.
```
Comment (praneesha) [inline learn/guides/visual-studio-code-extension/debugging.md:8]:
```suggestion
intro: The VS Code Ballerina language extension comes with in-built debugging capabilities and provides the same debugging experience as the conventional VS Code Debugging.
```
Shall we say `Visual Stucio Code` here as this is the intro?

File: learn/guides/visual-studio-code-extension/debugging.md [hunk 3]
```diff
@@ -29,11 +29,39 @@ redirect_from:
   - /learn/visual-studio-code-extension/debugging
 ---
 
->**Info:** For more information on debugging your code using VS Code, go to the [VS Code Documentation](https://code.visualstudio.com/docs/editor/debugging).
-
 ## Starting a Debug Session
 
-The VS Code Ballerina extension gives you the same debugging experience as the conventional VS Code Debugger. Follow the steps below to start a debug session.
+We can start a quick debug session instantly to debug a ballerina program with `CodeLens`. We also can start a debug session with configurations
```
Comment (praneesha) [inline learn/guides/visual-studio-code-extension/debugging.md:34]:
```suggestion
You can start a quick debug session instantly to debug a Ballerina program with `CodeLens`. Also, you can start a debug session with configurations
```
Please change `we` to `you` in all occurrences.

File: learn/guides/visual-studio-code-extension/debugging.md [hunk 4]
```diff
@@ -29,11 +29,39 @@ redirect_from:
...
+like program arguments and environment variables by adding them into the `launch.json` file.
+
+### Starting a Program Debug Session
+
+Ballerina extension provides multiple options to debug your Ballerina program and the easiest way will be using our context-aware `Debug` CodeLens support provided by the extension.
```
Comment (praneesha) [inline learn/guides/visual-studio-code-extension/debugging.md:39]:
```suggestion
The Ballerina extension provides multiple options to debug your Ballerina program and the easiest way will be using the context-aware `Debug` CodeLens support provided by the extension.
```

File: learn/guides/visual-studio-code-extension/debugging.md [hunk 5]
```diff
@@ -29,11 +29,39 @@ redirect_from:
...
+The extension can detect any program entry points (main function/service definition) on the fly, and the CodeLens will appear automatically.
```
Comment (praneesha) [inline learn/guides/visual-studio-code-extension/debugging.md:40]:
```suggestion
The extension can detect any program entry points (main function/service definition) on the fly and the CodeLens will appear automatically.  
```

File: learn/guides/visual-studio-code-extension/debugging.md [hunk 6]
```diff
@@ -29,11 +29,39 @@ redirect_from:
...
+
+![Start_Main Quick Debug Session](/learn/images/start-quick-main-debug-session.gif)
+
+<br/>
+
+1. Open the folder, which includes the Ballerina program you want to debug, and open the source file in the editor.
+
+2. Add the debug points you require by clicking in front of the line numbers of the file you want to debug.
+
+3. Click the `Debug` CodeLens which is just above the `main()` method.
```
Comment (praneesha) [inline learn/guides/visual-studio-code-extension/debugging.md:52]:
```suggestion
3. Click the `Debug` CodeLens, which is just above the `main()` method.
```

File: learn/guides/visual-studio-code-extension/debugging.md [hunk 7]
```diff
@@ -29,11 +29,39 @@ redirect_from:
...
+
+### Starting a Test Debug Session
+
+The Ballerina test functions can also be debugged using the codelens. The `debug` codelens will automatically appear on top of each Ballerina test function and the users are able to execute/debug only a selected test case by clicking on the corresponding codelens, as shown below.
```
Comment (praneesha) [inline learn/guides/visual-studio-code-extension/debugging.md:56]:
```suggestion
The Ballerina test functions can also be debugged using the codelens. The `debug` codelens will automatically appear on top of each Ballerina test function and you are able to execute/debug only a selected test case by clicking on the corresponding codelens as shown below.
```

File: learn/guides/visual-studio-code-extension/debugging.md [hunk 8]
```diff
@@ -89,9 +117,13 @@ Follow the steps below to start a test debug session.
 
 Follow the steps below to start a remote debug session.
 
-1. Select **Ballerina Remote** from the drop-down available in the upper left corner to start a remote debugging session.
+1. Create the `launch.json` configuration file if it is not created already. Refer [Debug Session with Configurations](#debug-session-with-configurations) section on how to create the `launch.json` file.
```
Comment (praneesha) [inline learn/guides/visual-studio-code-extension/debugging.md:120]:
```suggestion
1. Create the `launch.json` configuration file if it is not created already. For instructions on creating the `launch.json` file, see [Debug Session with Configurations](#debug-session-with-configurations).
```

File: learn/guides/visual-studio-code-extension/debugging.md [hunk 9]
```diff
@@ -89,9 +117,13 @@ Follow the steps below to start a test debug session.
...
 
-2. Open the Terminal and execute the Ballerina command, which you want to debug, out of the supported remote debugging commands below. 
+2. Open `launch.json` file and configure `debuggeeHost` and `debuggeePort` attributes under `Ballerina Remote` configurations section accordingly.
```
Comment (praneesha) [inline learn/guides/visual-studio-code-extension/debugging.md:122]:
```suggestion
2. Open the `launch.json` file and configure the `debuggeeHost` and `debuggeePort` attributes under the `Ballerina Remote` configurations section accordingly.
```

File: learn/guides/visual-studio-code-extension/debugging.md [hunk 10]
```diff
@@ -89,9 +117,13 @@ Follow the steps below to start a test debug session.
...
+
+3. After setting the remote debug configurations, select **Ballerina Remote** from the drop-down available in the upper left corner to start a remote debugging session.
+
+4. Open the Terminal and execute the Ballerina command, which you want to debug, out of the supported remote debugging commands below.
```
Comment (praneesha) [inline learn/guides/visual-studio-code-extension/debugging.md:126]:
```suggestion
4. Open the Terminal and execute the Ballerina command, which you want to debug out of the supported remote debugging commands below. 
```

File: learn/guides/visual-studio-code-extension/debugging.md [hunk 11]
```diff
@@ -117,33 +149,75 @@ Follow the steps below to start a remote debug session.
...
     You view the output in the **DEBUG CONSOLE**.
 
     ![Remote Debug](/learn/images/remote-debug.gif)
 
 <br/>
 
-### Running Without Debugging
+## Debug Configurations
+
+Ballerina debugger supports various debug configuration options via `launch.json` file. You can either add configurations to the existing `launch.json` file (which is located in your workspace root under the `.vscode` directory), or you can generate `launch.json` configurations file with default values by,
```
Comment (praneesha) [inline learn/guides/visual-studio-code-extension/debugging.md:162]:
```suggestion
The Ballerina debugger supports various debug configuration options via the `launch.json` file. 

>**Info:** You can either add configurations to the existing `launch.json` file (which is located in your workspace root under the `.vscode` directory), or you can generate the `launch.json` configurations file with default values by,
```
This sentence is incomplete.

File: learn/guides/visual-studio-code-extension/debugging.md [hunk 12]
```diff
@@ -117,33 +149,75 @@ Follow the steps below to start a remote debug session.
...
+
+1. Click the **Run and Debug** icon in the left menu or press the **Control + Shift + D** keys, to launch the Debugger view. (for Mac - **Command + Shift +D**).
```
Comment (praneesha) [inline learn/guides/visual-studio-code-extension/debugging.md:164]:
```suggestion
1. Click the **Run and Debug** icon in the left menu or press the **Control + Shift + D** keys to launch the Debugger view (for Mac - **Command + Shift +D**).
```

File: learn/guides/visual-studio-code-extension/debugging.md [hunk 13]
```diff
@@ -117,33 +149,75 @@ Follow the steps below to start a remote debug session.
...
 
-You view the program being executed in the **DEBUG CONSOLE** as shown below.
+Here are the default configurations generated for the Ballerina debugging:
 
-![Run Without Debugging](/learn/images/run-without-debugging.gif)
+![Debug Configurations](/learn/images/debug-configurations.png)
 
 <br/>
 
+>**Info:** You can debug a Ballerina program without generating the `launch.json` configurations file, but it is not possible to manage launch configurations and set up advanced debugging.
```
Comment (praneesha) [inline learn/guides/visual-studio-code-extension/debugging.md:182]:
```suggestion
>**Info:** You can debug a Ballerina program without generating the `launch.json` configurations file but it is not possible to manage launch configurations and set up advanced debugging.
```

File: learn/guides/visual-studio-code-extension/debugging.md [hunk 14]
```diff
@@ -117,33 +149,75 @@ Follow the steps below to start a remote debug session.
...
+
+### Ballerina launch.json attributes
```
Comment (praneesha) [inline learn/guides/visual-studio-code-extension/debugging.md:184]:
```suggestion
### Ballerina `launch.json` Attributes
```
Please make sure all titles follow the title case.

File: learn/guides/visual-studio-code-extension/debugging.md [hunk 15]
```diff
@@ -117,33 +149,75 @@ Follow the steps below to start a remote debug session.
...
+
+The auto-generated `launch.json` file consists of three main configurations, namely, `Ballerina Debug`, `Ballerina Test`, and `Ballerina Remote`. Each configuration supports different attributes, and those attributes can be identified with the help of IntelliSense suggestions (`⌃Space`).
```
Comment (praneesha) [inline learn/guides/visual-studio-code-extension/debugging.md:186]:
```suggestion
The auto-generated `launch.json` file consists of three main configurations namely, `Ballerina Debug`, `Ballerina Test`, and `Ballerina Remote`. Each configuration supports different attributes and those attributes can be identified with the help of IntelliSense suggestions (`⌃Space`).
```

File: learn/guides/visual-studio-code-extension/debugging.md [hunk 16]
```diff
@@ -117,33 +149,75 @@ Follow the steps below to start a remote debug session.
...
+
+The following attributes are mandatory for all configurations.
+
+- `name` - The reader-friendly name to appear in the Debug launch configuration dropdown.
+- `type` - The type of debugger to use for this launch configuration. The attribute value must be kept as `ballerina` for all Ballerina debugging configuration types.
+- `request` - The request type of this launch configuration. Currently, launch and attach are supported.
+
+The following attributes are supported for all Ballerina `launch` configurations.
+
+- `programArgs` - Any program arguments that are required to be passed into the `main` function of the Ballerina program to be launched, can be passed as a list of strings.
```
Comment (praneesha) [inline learn/guides/visual-studio-code-extension/debugging.md:196]:
```suggestion
- `programArgs` - Any program arguments that are required to be passed into the `main` function of the Ballerina program to be launched can be passed as a list of strings.
```

File: learn/guides/visual-studio-code-extension/debugging.md [hunk 17]
```diff
@@ -117,33 +149,75 @@ Follow the steps below to start a remote debug session.
...
+- `commandOptions` - If required, you can configure command options for the Ballerina program to be launched, as a list of strings. You can see the list of all the available command options by executing the following CLI commands in your terminal.
```
Comment (praneesha) [inline learn/guides/visual-studio-code-extension/debugging.md:197]:
```suggestion
- `commandOptions` - If required, you can configure command options for the Ballerina program to be launched as a list of strings. You can see the list of all the available command options by executing the following CLI commands in your terminal.
```

File: learn/guides/visual-studio-code-extension/debugging.md [hunk 18]
```diff
@@ -117,33 +149,75 @@ Follow the steps below to start a remote debug session.
...
+    - For `Ballerina Debug` configuration:
```
Comment (praneesha) [inline learn/guides/visual-studio-code-extension/debugging.md:198]:
```suggestion
    - For the `Ballerina Debug` configuration:
```

File: learn/guides/visual-studio-code-extension/debugging.md [hunk 19]
```diff
@@ -117,33 +149,75 @@ Follow the steps below to start a remote debug session.
...
+
+    ```bash
+    bal run --help
+    ```
+
+    - For `Ballerina Test` configuration:
```
Comment (praneesha) [inline learn/guides/visual-studio-code-extension/debugging.md:204]:
```suggestion
    - For the `Ballerina Test` configuration:
```

File: learn/guides/visual-studio-code-extension/debugging.md [hunk 20]
```diff
@@ -117,33 +149,75 @@ Follow the steps below to start a remote debug session.
...
+
+    ```bash
+    bal test --help
+    ```
+
+- `env` - Any environment variables you need to configure for the launching Ballerina program can be passed as a map of strings (name and value).
```
Comment (praneesha) [inline learn/guides/visual-studio-code-extension/debugging.md:210]:
```suggestion
- `env` - Any environment variables you need to configure for launching the Ballerina program can be passed as a map of strings (name and value).
```

File: learn/guides/visual-studio-code-extension/debugging.md [hunk 21]
```diff
@@ -117,33 +149,75 @@ Follow the steps below to start a remote debug session.
...
+- `debugTests` - Indicates whether to debug the tests for the given script.
+
+The following attributes are supported for all Ballerina `attach` configurations.
+
+- `debuggeeHost` - Host address of the remote process to be attached (If not specified, the default value will be the localhost(`127.0.0.1`)).
```
Comment (praneesha) [inline learn/guides/visual-studio-code-extension/debugging.md:215]:
```suggestion
- `debuggeeHost` - Host address of the remote process to be attached (if not specified, the default value will be the localhost(`127.0.0.1`)).
```

**Review by NipunaRanasinghe — COMMENTED** [review]
@praveennadarajah I'm seeing a single commit for each suggestion and 21 commits in total here. Shall we please avoid committing individual suggestions and squash them if possible?

File: learn/guides/visual-studio-code-extension/debugging.md (hunk 2 above)
Comment (praveennadarajah) [reply learn/guides/visual-studio-code-extension/debugging.md]: Fixed with https://github.com/ballerina-platform/ballerina-dev-website/pull/3385/commits/903828a68d9c603a8a3fec0a6a5ee665c1bcecac

**Comment by praveennadarajah** [comment]
> @praveennadarajah I'm seeing a single commit for each suggestion and 21 commits in total here. Shall we please avoid committing individual suggestions and squash them if possible?

@NipunaRanasinghe squashed all the review commits to one individual commit.

**Pull request merged**